service/src/main/kotlin/fi/espoo/evaka/messaging/bulletin/BulletinQueries.kt

```diff
@@ -126,28 +124,53 @@ fun Database.Transaction.sendBulletin(
     // language=sql
     val insertBulletinInstancesSql = """
         WITH children AS (
-            SELECT pl.child_id, br.id AS bulletin_receiver_id
+            SELECT child_id, bulletin_id FROM bulletin_receiver
+            WHERE
+            bulletin_id = :bulletinId 
+            AND unit_id IS NOT NULL 
+            AND group_id IS NOT NULL 
+            AND child_id IS NOT NULL
+            
+            UNION DISTINCT
+            
+            SELECT pl.child_id, br.bulletin_id
             FROM bulletin b
             JOIN bulletin_receiver br ON br.bulletin_id = b.id 
             JOIN daycare_group dg ON br.group_id = dg.id
             JOIN daycare_group_placement gpl ON dg.id = gpl.daycare_group_id AND daterange(gpl.start_date, gpl.end_date, '[]') @> :date
             JOIN placement pl ON gpl.daycare_placement_id = pl.id
-            WHERE b.id = :bulletinId
+            WHERE b.id = :bulletinId 
+            AND br.unit_id IS NOT NULL 
+            AND br.group_id IS NOT NULL 
+            AND br.child_id IS NULL
+            
+            UNION DISTINCT
+            
+            SELECT pl.child_id, br.bulletin_id
+            FROM bulletin b
+            JOIN bulletin_receiver br ON br.bulletin_id = b.id
+            JOIN daycare_group dg ON br.unit_id = dg.daycare_id
+            JOIN daycare_group_placement gpl ON dg.id = gpl.daycare_group_id AND daterange(gpl.start_date, gpl.end_date, '[]') @> :date
+            JOIN placement pl ON gpl.daycare_placement_id = pl.id
+            WHERE b.id = :bulletinId 
+            AND br.unit_id IS NOT NULL 
+            AND br.group_id IS NULL 
+            AND br.child_id IS NULL
```

Comment:
we are handling cases 
1. group not null, child not null
2. group not null, child is null
3. group is null, child is null
but what about the fourth possible permutation
4. group is null, child not null?

When does that happen, what does it mean, should it not be allowed? Depending on answers we should maybe either handle it here or prevent it with a database check constraint.

Reply:
let's discuss this through inbox